### ISSUE TITLE
Conda mangles channels with hyphens

### ISSUE THREAD

**Issue by bkreider** [conda/conda#5763]
Update: This is not related to hyphens, only substrings.

Seen on versions: 4.3.21, 4.3.23 (maybe more versions)

Conda mangles channels that are a substring match when using default_channels and channels.  The channel name "substring-channels" gets mangled into "substring/-channels".  It looks like it happens when there is a substring that matches.  

```
default_channels:
   - http://test/conda/substring
   - http://test/conda/default-channels

channels:
   - http://test/conda/substring-channels
   - http://test/conda/not_substring-channels
   - defaults
```

Here is the output of conda info
```
Current conda install:

               platform : linux-64
          conda version : 4.3.23
       conda is private : False
      conda-env version : 4.3.23
    conda-build version : 2.1.0
         python version : 2.7.13.final.0
       requests version : 2.12.4
       root environment : /home/ec2-user/miniconda2  (writable)
    default environment : /home/ec2-user/miniconda2
       envs directories : /home/ec2-user/miniconda2/envs
                          /home/ec2-user/.conda/envs
          package cache : /home/ec2-user/miniconda2/pkgs
                          /home/ec2-user/.conda/pkgs
           channel URLs : http://test/conda/substring/-channels/linux-64   <<<<<<<<<<<<<<====
                          http://test/conda/substring/-channels/noarch   <<<<<<<<<<<<<<====
                          http://test/conda/not_substring-channels/linux-64
                          http://test/conda/not_substring-channels/noarch
                          http://test/conda/substring/linux-64
                          http://test/conda/substring/noarch
                          http://test/conda/default-channels/linux-64
                          http://test/conda/default-channels/noarch
            config file : /home/ec2-user/miniconda2/.condarc
             netrc file : None
           offline mode : False
             user-agent : conda/4.3.23 requests/2.12.4 CPython/2.7.13 Linux/4.4.11-23.53.amzn1.x86_64 / glibc/2.17    
                UID:GID : 500:500
```

Also `conda info --unsafe-channels` seems to give weird output.  It doesn't expand "defaults".

**Comment by bkreider** [comment]
I've updated this to say that it is substrings between channels and default_channels that causes this bugs, not hyphens.  Here is a complete example.  Substrings from default_channels cause problems with "channels".

```
default_channels:
   - http://test/conda/substring  <<<<< PROBLEM
   - http://test/conda/default-channels
   - http://test/conda/sub            <<<<< PROBLEM
   - http://test/conda/sub2          <<<<< Not a problem
   - http://test/conda/sub2sub2  <<<<< not a problem

channels:
   - http://test/conda/substring-channels
   - http://test/conda/not_substring-channels
   - http://test/conda/subsub
   - defaults
```

**Comment by bkreider** [comment]
It's this bit of logic:  https://github.com/conda/conda/blob/master/conda/models/channel.py#L397-L404

I don't know why this does this, but that splits off anything that is a substring and joins it as a separate path component. 
```
    # Step 4. custom_channels matches
    for name, channel in sorted(context.custom_channels.items(), reverse=True,
                                key=lambda x: len(x[0])):
        that_test_url = join_url(channel.location, channel.name)
        if test_url.startswith(that_test_url):
            subname = test_url.replace(that_test_url, '', 1).strip('/')
            return (channel.location, join_url(channel.name, subname), scheme,
                    channel.auth, channel.token)
```

**Comment by kalefranz** [comment]
@bkreider have the fix in #5764. It'll go into 4.3.x.
